### PR TITLE
Config features and better `unstack_dates`

### PR DESCRIPTION
**Pull request by aulemahal**
### Pull Request Checklist:
- [ ] This PR addresses an already opened issue (for bug fixes / features)
    - This PR fixes #xyz
- [x] (If applicable) Documentation has been added / updated (for bug fixes / features)
- [ ] HISTORY.rst has been updated (with summary of main changes)
  - [ ] Link to issue (:issue:`number`) and pull request (:pull:`number`) has been added

### What kind of change does this PR introduce?

* Add a `CONFIG.set` method.  Each time an item of  `CONFIG` is accessed, a copy is returned, in order to avoid errors. This adds back the possibility to edit the configuration programmatically.
* Larger scope of `load_config` so it accepts `key=value` entries.  This is meant to make it easier to parse from the CLI.
* Rewrite of a condition in `extract_dataset`, the changed line is supposed to be equivalent to the old one, given how those variables are set above, but it also avoids a bug when a `var_name` is not given in `variables_and_freq`.
* `to_dataset_dict` only prints a progress bar if there are more than 1 dataset. Avoids useless output in trivial cases.

#### Rewrite of `unstack_dates`
Instead of using xarray,s unstack,  I try here to perform a `reshape` operation. This is much more efficient with dask, but it did necessitate heavy changes. 

* This new version will fail if `xr.infer_freq(ds)` fails. I assumed this is almost never the case for the use we make of this function.
* The `seasons` argument is now a dictionary with the month number as keys, not a "%m-%d" date. The function currently only works for coarser-than-month frequencoes, so this simply made the code easier to write. Also, even if we expand to accept `daily` freqs, the  `seasons` arg wouldn't have any meaning I think (the new dim would be dayofyear anyway, no ?).
* Reshape is not a xarray method, the data's one is used instead, necessitating some DataArray/Dataset manipulation.
* And most importantest: to be able to perform a `reshape` on dask arrays, the chunks are realigned using `flox` beforehand, ensuring the chunks frontiers fall on yearly timesteps. 
* Flox as a new hard dependency.

### Does this PR introduce a breaking change?
`unstack_dates` has lost some flexibility : data must pass `xr.infer_freq`.
`flox` as a dependency.
No more progress bar printed by `extract_dataset` for single datasets.

### Other information:
I struggled to run the notebooks. But I'll try to do it soon, as a test for breaking changes.

### CODE REVIEW

**Review by juliettelavoie — COMMENTED** [review]
`unstack_dates` is used in `extract.subset_warming_level`. It would be good to test that this still works in the warming levels notebook.

Also, I feel like the flox reshape trick makes the code more opaque, but it is probably worth it if allows it to be quicker...

**Comment by aulemahal** [comment]
@juliettelavoie For illustration. I take all QS-DEC indicators of ScenGen and construct an ensemble (dims: `(experiment: 2, realization: 11, time: 605, lat: 320, lon: 416)`). I then unstack the seasons and perform a spatial mean over 33 regions + percentiles over realizations.

The "stacked" dataset has 71508 tasks (`len(ds.__dask_graph__().keys())`).

Before:
`unstack_seasons` took 1470 ms and added 118 560 tasks. The final compute took 12 min 19s.

After (this branch):
`unstack_seasons` took 381 ms (0.25x) add added 24 730 tasks (0.21x). The final compute took 5 min 9s (0.42x).

All that said, I realize the code is more opaque... I can try to make it clearer, but I felt the performance boost was worth it.

File: xscen/config.py [hunk 1]
```diff
@@ -125,31 +127,60 @@ def args_as_str(*args: Tuple[Any, ...]) -> Tuple[str, ...]:
     return tuple(new_args)
 
 
-def load_config(*files, reset=False, verbose=False):
-    """Load configuration from given files (in order, the last has priority).
+def load_config(*elements, reset=False, verbose=False):
+    """Load configuration from given files or key=value pairs (in order, the last has priority).
```
[the remark this answers is not in the document]
Comment (aulemahal) [reply xscen/config.py:131]: "key=value" pairs have priority over files. I'll rephrase this.

File: xscen/config.py [hunk 2]
```diff
@@ -125,31 +127,60 @@ def args_as_str(*args: Tuple[Any, ...]) -> Tuple[str, ...]:
...
+        If a a directory is passed, all `.yml` files of this directory are added, in alphabetical order.
+        If a "key=value" string, "key" is a dotted name and value will be evaluated if possible.
+    reset: bool
+        If True, the current config is erased before loading files.
+    verbose: bool
+        if True, each element triggers a INFO log line.
+
+    Example
+    -------
+    >>> load_config("my_config.yml", "config_dir/", "xarray.keep_attrs=True")
```
[the remark this answers is not in the document]
Comment (aulemahal) [reply xscen/config.py:154]: I  wanted to find a "key=value" pair that was understandable for the example. Indeed, this is not a very realist use case...

File: xscen/utils.py [hunk 3]
```diff
@@ -719,82 +721,156 @@ def publish_release_notes(
...
+    winter_starts_year: bool = False,
 ):
     """Unstack a multi-season timeseries into a yearly axis and a season one.
 
     Parameters
     ----------
     ds: xr.Dataset or DataArray
       The xarray object with a "time" coordinate.
+      Only supports monthly or coarser frequencies.
+      The time axis must be complete and regular (`xr.infer_freq(ds.time)` doesn't fail).
```
[the remark this answers is not in the document]
Comment (aulemahal) [reply xscen/utils.py:735]: I didn't want to add to many pre-processing steps to this as I thought they could be taken care of outside, with project-specific solutions.

For example, your 2. is easily avoidable with: `ds = ds.resample(time='MS').mean()`. 

I didn't think of 1. indeed... 

A solution would be to accept a  `freq` argument to use instead if `infer_freq` fails. Except for the len-2 case, we could then repair the time coordinate in the same manner as in my example, but with more safeguards.

File: xscen/utils.py [hunk 4]
```diff
@@ -719,82 +721,156 @@ def publish_release_notes(
...
 
     Notes
     -----
-    When `seasons` is None, :py:func:`xarray.infer_freq` is called and its output determines the new coordinate:
+    When `season` is None, the inferred frequency determines the new coordinate:
 
     - For MS, the coordinates are the month abbreviations in english (JAN, FEB, etc.)
     - For ?QS-? and other ?MS frequencies, the coordinates are the initials of the months in each season.
-      Ex: QS-DEC : DJF, MAM, JJA, SON.
+      Ex: QS-DEC (with winter_starts_year=True) : DJF, MAM, JJA, SON.
```
[the remark this answers is not in the document]
Comment (aulemahal) [reply xscen/utils.py:756]: If `winter_starts_year` is false. Then, `MAM` is the first season of the year and `DJF` is the last. The new `season` coordinate would then be: `MAM, JJA, SON, DJF`.

File: xscen/utils.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (aulemahal) [reply xscen/utils.py:735]: It still feel this is kinda out-of-scope for the function, but it is also easy to add, or to simply explain.

Do you prefer:
```python3
out = unstack_seasons(ds)
# ValueError : The data must have a clean time coordinate. If you know the data's frequency, please pass `ds.resample(time=freq).first()` to pad missing dates and reset the time coordinate.
out = unstack_seasons(ds.resample(time='MS').first())
```
OR
```python3
out = unstack_seasons(ds, freq='MS')
```
Where the first lines of the function are:
```python3
if freq is None:
    freq = xr.infer_freq(ds.time)
else:
    ds = ds.resample(time=freq).first()
```

[the remark this answers is not in the document]
Comment (juliettelavoie) [reply xscen/utils.py:736]: I think the explanation in the Error is probably enough.

File: xscen/utils.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (juliettelavoie) [reply xscen/utils.py:758]: ah yes

**Comment by aulemahal** [comment]
Latest pushes fix my fix and all notebooks should run. As per `produce_horizon`, its use of `unstack_dates` should not be affected.

**Review by juliettelavoie — APPROVED** [review]

**Review by RondeauG — APPROVED** [review]

**Pull request merged**